modules/foreign-state-oracle/src/lib.rs

```diff
@@ -0,0 +1,326 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Foreign State Oracle.
+//! This module provide interface for other internal modules to create a QueryRequest that
+//! requires Oracles to verify external states (such as states on the Relay Chain).
+//! Basic workflow:
+//! 1. Internal module will create a QueryRequest, providing a dispatchable "callback"
+//! 2. External oracles will then verify states and provide feedback
+//! 3. Feedback are tallied, depending on how `OracleOrigin` is configured
+//! 4. This module will dispatch the callback, injecting any response data into the Origin.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo},
+	pallet_prelude::*,
+	require_transactional,
+	traits::{Currency, ExistenceRequirement},
+	transactional, PalletId,
+};
+use frame_system::pallet_prelude::*;
+use module_support::ForeignChainStateQuery;
+use primitives::Balance;
+use sp_runtime::{
+	traits::{AccountIdConversion, BlockNumberProvider, One, Saturating},
+	ArithmeticError, Permill,
+};
+use sp_std::{ops::Mul, prelude::Vec};
+
+mod mock;
+mod tests;
+pub mod weights;
+
+pub use module::*;
+pub use weights::WeightInfo;
+
+// Unique Identifier for each query
+pub type QueryIndex = u64;
+
+// Origin with arbitrary bytes included
+#[derive(PartialEq, Eq, Clone, RuntimeDebug, Encode, Decode, TypeInfo)]
+pub struct RawOrigin {
+	data: Vec<u8>,
+}
+
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct ForeignQueryRequest<Call, BlockNumber> {
+	// Call to be dispatched by oracle
+	dispatchable_call: Call,
+	// Blocknumber at which call will be expired
+	expiry: BlockNumber,
+}
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	pub type ForeignQueryRequestOf<T> =
+		ForeignQueryRequest<<T as Config>::DispatchableCall, <T as frame_system::Config>::BlockNumber>;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The outer origin type.
+		type Origin: From<RawOrigin>;
+
+		/// Weight Info
+		type WeightInfo: WeightInfo;
+
+		/// Currency for query payments
+		type Currency: Currency<Self::AccountId, Balance = Balance>;
+
+		/// The foreign state oracle module id, keeps expired queries deposits
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Fee to be paid to oracles for servicing query
+		#[pallet::constant]
+		type QueryFee: Get<Balance>;
+
+		/// Fee for cancelling query
+		#[pallet::constant]
+		type CancelFee: Get<Balance>;
+
+		/// Timeout for query requests
+		#[pallet::constant]
+		type DefaultQueryDuration: Get<Self::BlockNumber>;
```

Comment:
I don't think we should have this. We should allow query that never expires.

Reply:
None now represents query that never expires